packages/lesswrong/server/languageModels/autoTagCallbacks.ts

```diff
@@ -0,0 +1,246 @@
+  const markdownPostBody = postHtml ? dataToMarkdown(preprocessHtml(postHtml), "html") : "";
+  return markdownPostBody;
+}
+
+export type PostBodyCache = {preprocessedBody: Record<string,string>}
+export function generatePostBodyCache(posts: DbPost[]): PostBodyCache {
+  const result = {preprocessedBody: {}};
+  for (let post of posts) {
+    result.preprocessedBody[post._id] = preprocessPostBody(post);
+  }
+  return result;
+}
+
+export async function checkTags(post: DbPost, tags: DbTag[], openAIApi: OpenAIApi) {
+  const template = await wikiSlugToTemplate("lm-config-autotag");
+  
+  let tagsApplied = {};
+  
+  for (let tag of tags) {
+    const languageModelResult = await openAIApi.createCompletion({
+      model: tag.autoTagModel,
+      prompt: await postToPrompt({template, post, promptSuffix: tag.autoTagPrompt}),
+      max_tokens: 1,
+    });
+    const completion = languageModelResult.data.choices[0].text!;
+    const hasTag = (completion.trim().toLowerCase() === "yes");
+    tagsApplied[tag.slug] = hasTag;
+  }
+  
+  return tagsApplied;
+}
+
+
+async function getTagBotAccount(context: ResolverContext): Promise<DbUser|null> {
+  const accountSlug = tagBotAccountSlug.get();
+  if (!accountSlug) return null;
+  const account = await Users.findOne({slug: accountSlug});
+  if (!account) return null;
+  return account;
+}
+
+let tagBotUserIdCache: {id:string|null}|null = null;
+export async function getTagBotUserId(context: ResolverContext): Promise<string|null> {
+  if (!tagBotUserIdCache) {
+    const tagBotAccount = await getTagBotAccount(context);
+    tagBotUserIdCache = {id: tagBotAccount?._id ?? null};
+  }
+  return tagBotUserIdCache.id;
+}
+
+export async function getAutoAppliedTags(): Promise<DbTag[]> {
+  return await Tags.find({ autoTagPrompt: {$exists: true, $ne: ""} }).fetch();
+}
+
+async function autoApplyTagsTo(post: DbPost, context: ResolverContext): Promise<void> {
+  const api = await getOpenAI();
+  if (!api) {
+    //eslint-disable-next-line no-console
+    console.log("Skipping autotagging (API not configured)");
+    return;
+  }
+  const tagBot = await getTagBotAccount(context);
+  if (!tagBot) {
+    //eslint-disable-next-line no-console
+    console.log("Skipping autotagging (no tag-bot account)");
+    return;
+  }
+  
+  const tags = await getAutoAppliedTags();
+  const tagsApplied = await checkTags(post, tags, api);
+  
+  //eslint-disable-next-line no-console
+  console.log(`Auto-applying tags to post ${post.title} (${post._id}): ${JSON.stringify(tagsApplied)}`);
+  
+  for (let tag of tags) {
+    if (tagsApplied[tag.slug]) {
+      await addOrUpvoteTag({
+        tagId: tag._id,
+        postId: post._id,
+        currentUser: tagBot,
+        context,
+      });
+    }
+  }
+}
+
+getCollectionHooks("Posts").updateAsync.add(async ({oldDocument, newDocument, context}) => {
+  if (oldDocument.draft && !newDocument.draft) {
+    // Post was undrafted
+    void autoApplyTagsTo(newDocument, context);
+  }
+})
```

Comment:
Will this avoid auto-applying tags multiple times, e.g. in the case of a post being set back to draft after being posted, then undrafted again, because the `tagBot` can only upvote a tag once?

Reply:
Each time it's undrafted, every tag will be rechecked. It can't apply the same tag twice because the bot account can only upvote once, but I guess you could trick the tagbot account into applying all the tags by replacing it with different post bodies each time you undraft it.

Reply:
Eh, that seems fine